pkg/scheduler/util/non_zero.go

```diff
@@ -34,45 +35,59 @@ const (
+	DefaultEphemeralStorage int64 = 0 // 0 Gi
 )
 
 // GetNonzeroRequests returns the default cpu and memory resource request if none is found or
 // what is provided on the request.
-func GetNonzeroRequests(requests *v1.ResourceList) (int64, int64) {
-	return GetNonzeroRequestForResource(v1.ResourceCPU, requests),
-		GetNonzeroRequestForResource(v1.ResourceMemory, requests)
+func GetNonzeroRequests(pod *v1.Pod) (int64, int64) {
+	return GetNonzeroRequestForResource(pod, v1.ResourceCPU, true),
+		GetNonzeroRequestForResource(pod, v1.ResourceMemory, true)
 }
 
 // GetNonzeroRequestForResource returns the default resource request if none is found or
 // what is provided on the request.
-func GetNonzeroRequestForResource(resource v1.ResourceName, requests *v1.ResourceList) int64 {
-	switch resource {
-	case v1.ResourceCPU:
-		// Override if un-set, but not if explicitly set to zero
-		if _, found := (*requests)[v1.ResourceCPU]; !found {
-			return DefaultMilliCPURequest
-		}
-		return requests.Cpu().MilliValue()
-	case v1.ResourceMemory:
-		// Override if un-set, but not if explicitly set to zero
-		if _, found := (*requests)[v1.ResourceMemory]; !found {
-			return DefaultMemoryRequest
-		}
-		return requests.Memory().Value()
-	case v1.ResourceEphemeralStorage:
-		quantity, found := (*requests)[v1.ResourceEphemeralStorage]
-		if !found {
-			return 0
+func GetNonzeroRequestForResource(pod *v1.Pod, resource v1.ResourceName, enforcePodOverhead bool) int64 {
+	opts := resourcehelper.WithEnforcePodOverhead(enforcePodOverhead)
+	podRequest := resourcehelper.GetResourceRequest(pod, resource, opts)
+
+	// If Overhead is being utilized, add to the total requests for the pod
+	for i := range pod.Spec.Containers {
+		container := &pod.Spec.Containers[i]
+		requests := container.Resources.Requests
+
+		// If the pod has an explicit resource request(includes initContainers),
+		// we shouldn't assign a default value to it.
+		if podRequest != 0 {
+			break
 		}
-		return quantity.Value()
-	default:
-		if v1helper.IsScalarResourceName(resource) {
-			quantity, found := (*requests)[resource]
-			if !found {
-				return 0
+
+		switch resource {
+		case v1.ResourceCPU:
+			// Override if un-set, but not if explicitly set to zero
+			if _, found := (requests)[v1.ResourceCPU]; !found {
+				podRequest = DefaultMilliCPURequest
+			}
+		case v1.ResourceMemory:
+			// Override if un-set, but not if explicitly set to zero
+			if _, found := (requests)[v1.ResourceMemory]; !found {
+				podRequest = DefaultMemoryRequest
+			}
+		case v1.ResourceEphemeralStorage:
+			if _, found := (requests)[v1.ResourceEphemeralStorage]; !found {
+				podRequest = DefaultEphemeralStorage
+			}
+		default:
+			if v1helper.IsScalarResourceName(resource) {
+				quantity, found := (requests)[resource]
+				if !found {
+					podRequest = 0
+				}
+				podRequest = quantity.Value()
 			}
-			return quantity.Value()
 		}
 	}
-	return 0
+
+	return podRequest
```

Comment:
Given that we have the single source of truth: `resourcehelper.GetResourceRequest`, and overhead has been covered. I suggest we check if the summed request is 0 or not (which includes podOverhead request):

```go
	if podRequest != 0 {
		return podRequest
	}
	switch resource {
	case v1.ResourceCPU:
		return DefaultMilliCPURequest
	case v1.ResourceMemory:
		return DefaultMemoryRequest
	case v1.ResourceEphemeralStorage:
		return DefaultEphemeralStorage
	default:
		// Don't give default value for scalar resource.
	}

	return 0
```

**Note** that this would change the old logic, so some use cases have to be updated.